torchtext/modules/multiheadattention.py

```diff
@@ -0,0 +1,205 @@
+            >>> query = torch.rand((21, bsz, embed_dim))
+            >>> key = value = torch.rand((16, bsz, embed_dim))
+            >>> attn_output, attn_weights = MHA(query, key, value)
+            >>> print(attn_output.shape)
+            >>> torch.Size([21, 64, 10])
+        """
+        super(MultiheadAttentionContainer, self).__init__()
+        self.nhead = nhead
+        self.in_proj_container = in_proj_container
+        self.attention_layer = attention_layer
+        self.out_proj = out_proj
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor,
+                attn_mask: Optional[torch.Tensor] = None,
+                bias_k: Optional[torch.Tensor] = None,
+                bias_v: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+
+        Args:
+            query, key, value (Tensor): map a query and a set of key-value pairs to an output.
+                See "Attention Is All You Need" for more details.
+            attn_mask (BoolTensor, optional): 3D mask that prevents attention to certain positions.
+            bias_k and bias_v: (Tensor, optional): one more key and value sequence to be added at
+                sequence dim (dim=-3). Those are used for incremental decoding. Users should provide
+                non-None to both arguments in order to activate them.
+
+        Shape:
+            - Inputs:
+            - query: :math:`(L, N, E)`
+            - key: :math:`(S, N, E)`
+            - value: :math:`(S, N, E)`
+            - attn_mask: :math:`(N * H, L, S)`, positions with ``True`` are not allowed to attend
+                while ``False`` values will be unchanged.
+            - bias_k and bias_v:bias: :math:`(1, N * H, E / H)`
+
+            - Outputs:
+            - attn_output: :math:`(L, N, E)`
+            - attn_output_weights: :math:`(N*num_heads, L, S)`
+
+            where where L is the target length, S is the sequence length, H is the number of attention heads,
+                N is the batch size, and E is the embedding dimension.
+        """
+        tgt_len, src_len, bsz, embed_dim = query.size(-3), key.size(-3), query.size(-2), query.size(-1)
+        q, k, v = self.in_proj_container(query, key, value)
+        assert q.size(-1) % self.nhead == 0, "query's embed_dim must be divisible by the number of heads"
+        head_dim = q.size(-1) // self.nhead
+        q = q.reshape(tgt_len, bsz * self.nhead, head_dim)
+
+        assert k.size(-1) % self.nhead == 0, "key's embed_dim must be divisible by the number of heads"
+        head_dim = k.size(-1) // self.nhead
+        k = k.reshape(src_len, bsz * self.nhead, head_dim)
+
+        assert v.size(-1) % self.nhead == 0, "value's embed_dim must be divisible by the number of heads"
+        head_dim = v.size(-1) // self.nhead
+        v = v.reshape(src_len, bsz * self.nhead, head_dim)
+
+        attn_output, attn_output_weights = self.attention_layer(q, k, v, attn_mask=attn_mask,
```

Comment:
It seems that for this container in particular there are no assumptions made on the dtype of attn_mask. I think we can relax that constraint. It stems from the fact that ScaledDotProduct needs a BoolTensor as a mask, but not for the container.